server/routes/room-routes.js

```diff
@@ -153,33 +199,134 @@ router.get(
     const roomId = await ensureRoomJoined(matrixAccessToken, roomIdOrAlias, req.query.via);
 
-    let eventIdForTimestamp;
-    let originServerTs;
+    let eventIdForClosestEvent;
+    let tsForClosestEvent;
+    let newOriginServerTs;
+    let preferredPrecision = null;
     try {
-      // Find the closest day to today with messages
-      ({ eventId: eventIdForTimestamp, originServerTs } = await timestampToEvent({
-        accessToken: matrixAccessToken,
-        roomId,
-        ts: ts,
-        direction: dir,
-      }));
+      const archiveMessageLimit = config.get('archiveMessageLimit');
 
+      console.log('jumping from ts', ts, new Date(ts).toISOString());
+
+      // Find the closest event to the given timestamp
+      ({ eventId: eventIdForClosestEvent, originServerTs: tsForClosestEvent } =
+        await timestampToEvent({
+          accessToken: matrixAccessToken,
+          roomId,
+          ts: ts,
+          direction: dir,
+        }));
+      console.log(
+        'timestamp_to_event found',
+        eventIdForClosestEvent,
+        tsForClosestEvent,
+        new Date(tsForClosestEvent).toISOString()
+      );
+
+      // Based on what we found was the closest, figure out the URL that will represent
+      // the next chunk in the desired direction.
+      // ==============================
+      //
+      // Since a given room archive URL represents the end of the day/time-period
+      // looking backward (scroll is also anchored to the bottom), we just need to get
+      // the user to the previous time-period.
+      //
+      // We are trying to avoid sending the user to the same time period they were just
+      // viewing. i.e, if they were visiting `/2020/01/02T23:59:59`, which had more
+      // messages than we could display in that day, jumping backwards from the earliest
+      // displayed event from `T16:00:00` would still give us the same day `/2020/01/02`
+      // and we want to redirect them to previous chunk from that same day, like
+      // `/2020/01/02T16:00:00`
+      if (dir === 'b') {
+        const dateOfClosestEvent = new Date(tsForClosestEvent);
+
+        const fromSameDay = areTimestampsFromSameDay(currentRangeEndTs, tsForClosestEvent);
+        const fromSameHour = areTimestampsFromSameHour(currentRangeEndTs, tsForClosestEvent);
+        const fromSameMinute = areTimestampsFromSameMinute(currentRangeEndTs, tsForClosestEvent);
+        const fromSameSecond = areTimestampsFromSameSecond(currentRangeEndTs, tsForClosestEvent);
+
+        console.log('fromSameDay', fromSameDay);
+        console.log('fromSameHour', fromSameHour);
+        console.log('fromSameMinute', fromSameMinute);
+        console.log('fromSameSecond', fromSameSecond);
+
+        // The closest event is from the same second we tried to jump from. Since we
+        // can't represent something smaller than a second in the URL (we could do MS
+        // but it's a concious choice to make the URL cleaner), we will need to just
```

Comment:
I think we could let the URL support `ms` as long as we got rid of the `/date/2022/11/16T23:59:59` default. That should be equivalent to `/date/2022/11/16` anyway

Reply:
> We can also assume that `T23:59:59` is equivalent to `T23:59:59.999` (instead of `T23:59:59.000`)

This is flawed and probably not good to do because then how do we represent the `T23:59:59.000` case.

Regardless, we should be able to coalesce to `/date/2022/11/16` somehow

Reply:
👍 Removed the `T23:59:59.000` stuff